stratum/hal/lib/barefoot/p4runtime_bfrt_translator.cc

```diff
@@ -0,0 +1,742 @@
+                           translated_entity.direct_counter_entry(), to_sdk));
+      break;
+    }
+    case ::p4::v1::Entity::kCounterEntry: {
+      ASSIGN_OR_RETURN(
+          *translated_entity.mutable_counter_entry(),
+          TranslateCounterEntry(translated_entity.counter_entry(), to_sdk));
+      break;
+    }
+    case ::p4::v1::Entity::kRegisterEntry: {
+      ASSIGN_OR_RETURN(
+          *translated_entity.mutable_register_entry(),
+          TranslateRegisterEntry(translated_entity.register_entry(), to_sdk));
+      break;
+    }
+    case ::p4::v1::Entity::kDirectMeterEntry: {
+      ASSIGN_OR_RETURN(*translated_entity.mutable_direct_meter_entry(),
+                       TranslateDirectMeterEntry(
+                           translated_entity.direct_meter_entry(), to_sdk));
+      break;
+    }
+    case ::p4::v1::Entity::kMeterEntry: {
+      ASSIGN_OR_RETURN(
+          *translated_entity.mutable_meter_entry(),
+          TranslateMeterEntry(translated_entity.meter_entry(), to_sdk));
+      break;
+    }
+    default:
+      break;
+  }
+  return translated_entity;
+}
+
+::util::StatusOr<::p4::v1::TableEntry>
+P4RuntimeBfrtTranslator::TranslateTableEntry(const ::p4::v1::TableEntry& entry,
+                                             bool to_sdk) {
+  ::p4::v1::TableEntry translated_entry(entry);
+  const auto& table_id = translated_entry.table_id();
+  if (table_to_field_to_type_uri_.count(table_id) &&
+      table_to_field_to_bit_width_.count(table_id)) {
+    for (::p4::v1::FieldMatch& field_match :
+         *translated_entry.mutable_match()) {
+      const auto& field_id = field_match.field_id();
+      std::string* uri =
+          gtl::FindOrNull(table_to_field_to_type_uri_[table_id], field_id);
+
+      int32 from_bit_width = 0;
+      int32 to_bit_width = 0;
+      if (to_sdk && uri) {
+        from_bit_width = gtl::FindWithDefault(
+            table_to_field_to_bit_width_[table_id], field_id, 0);
+        to_bit_width = gtl::FindWithDefault(kUriToBitWidth, *uri, 0);
+      } else if (uri) {
+        from_bit_width = gtl::FindWithDefault(kUriToBitWidth, *uri, 0);
+        to_bit_width = gtl::FindWithDefault(
+            table_to_field_to_bit_width_[table_id], field_id, 0);
+      }
+      if (uri && from_bit_width && to_bit_width) {
+        switch (field_match.field_match_type_case()) {
+          case ::p4::v1::FieldMatch::kExact: {
+            ASSIGN_OR_RETURN(const std::string& new_val,
+                             TranslateValue(field_match.exact().value(), *uri,
+                                            to_sdk, to_bit_width));
+            field_match.mutable_exact()->set_value(new_val);
+            break;
+          }
+          case ::p4::v1::FieldMatch::kTernary: {
+            // We only allow the "exact" type of ternary match, which means
+            // all bits from mask must be one.
+            std::string all_one = std::string(from_bit_width / 8, '\xff');
```

Comment:
Is truncation intended here? `7 / 8 = 0`

```suggestion
            std::string all_one = std::string(NumBitsToNumBytes(from_bit_width), '\xff');
```

ok, I see that the next line inserts the leading byte. I noticed that `RangeDefaultHigh` does the same thing, chance of code reuse?

Reply:
I add a new function `MaxValueOfBits` which will be used by `RangeDefaultHigh` and here.